pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+		Salt:   salt,
+		crypto: crypto,
+	}
+}
+
+func New() Crypto {
+	cryptoBackend = defaultCryptoBackend{
+		symKeys:     make(map[string][]byte),
+		privateKeys: make(map[string]*ecdsa.PrivateKey),
+	}
+	return &cryptoBackend
+}
+
+func NewCryptoUtils() CryptoUtils {
+	if cryptoBackend.privateKeys == nil {
+		New()
+	}
+	return &cryptoBackend
+}
+
+// == Message Crypto ==
+
+func (crypto *defaultCryptoBackend) WrapMessage(payload []byte, params *WrapParams) (data []byte, err error) {
+	var padding []byte
+	if createPadding {
+		padding, err = crypto.generateSecureRandomData(defaultPaddingByteSize)
+		if err != nil {
+			return
+		}
+	} else {
+		padding = make([]byte, 0)
+	}
+	// Message structure is flags+PayloadSize+Payload+padding+signature
+	rawBytes := make([]byte, 1,
+		flagsLength+payloadSizeFieldMaxSize+len(payload)+len(padding)+signatureLength+padSizeLimit)
+	// set flags byte
+	rawBytes[0] = 0 // set all the flags to zero
+	// add payloadSizeField
+	rawBytes = crypto.addPayloadSizeField(rawBytes, payload)
+	// add payload
+	rawBytes = append(rawBytes, payload...)
+	// add padding
+	rawBytes = append(rawBytes, padding...)
+	// sign
+	if params.Src != nil {
+		if rawBytes, err = crypto.sign(rawBytes, params.Src); err != nil {
+			return
+		}
+	}
+	// encrypt
+	if params.Dst != nil {
+		rawBytes, err = crypto.encryptAsymmetric(rawBytes, params.Dst)
+	} else if params.KeySym != nil {
+		rawBytes, err = crypto.encryptSymmetric(rawBytes, params.KeySym)
+	} else {
+		err = errNeitherSymNorAsymKeysProv
+	}
+	if err != nil {
+		return
+	}
+
+	data = rawBytes
+	return
+}
+
+func (crypto *defaultCryptoBackend) UnWrapSymmetric(encryptedData, key []byte) (ReceivedMessage, error) {
+	decrypted, salt, err := crypto.decryptSymmetric(encryptedData, key)
+	if err != nil {
+		return nil, err
+	}
+	msg := newReceivedMessage(decrypted, salt, crypto)
+	return msg, err
+}
+
+func (crypto *defaultCryptoBackend) UnWrapAsymmetric(encryptedData []byte, key *ecdsa.PrivateKey) (ReceivedMessage, error) {
+	decrypted, err := crypto.decryptAsymmetric(encryptedData, key)
+	switch err {
+	case nil:
+		message := newReceivedMessage(decrypted, nil, crypto)
+		return message, nil
+	case ecies.ErrInvalidPublicKey: // addressed to somebody else
+		return nil, err
+	default:
+		return nil, fmt.Errorf("unable to open envelope, decrypt failed: %v", err)
+	}
+}
+
+func (crypto *defaultCryptoBackend) addPayloadSizeField(rawBytes rawMessage, payload []byte) rawMessage {
+	fieldSize := getSizeOfPayloadSizeField(payload)
+	field := make([]byte, 4)
+	binary.LittleEndian.PutUint32(field, uint32(len(payload)))
+	field = field[:fieldSize]
+	rawBytes = append(rawBytes, field...)
+	rawBytes[0] |= byte(fieldSize)
+	return rawBytes
+}
+
+// appendPadding appends the padding specified in params.
+// If no padding is provided in params, then random padding is generated.
+func (crypto *defaultCryptoBackend) appendPadding(rawBytes, payload []byte, src *ecdsa.PrivateKey) (rawMessage, error) {
```

Comment:
rename to `pad`?

Reply:
passing around private key is unnecessary and less secure

Reply:
You are right, changed param to a boolean `signed`.